File: src/main/java/com/mattchowning/file_read_write/server/model/OAuthTokenMap.java
package com.mattchowning.file_read_write.server.model;

import java.util.HashSet;
import java.util.Set;
import java.util.function.Predicate;

public class OAuthTokenMap {

    private Set<OAuthToken> issuedTokens = new HashSet<>();

    public boolean add(OAuthToken oAuthToken) {
        boolean alreadyContainsToken = issuedTokens.stream()
                                                   .anyMatch(issuedToken -> oAuthToken == issuedToken);
        if (alreadyContainsToken) {
            System.err.println("Error: token already added to TokenMap");
            return false;
        } else {
            issuedTokens.add(oAuthToken);
            return true;
        }
    }

    public boolean containsAccessToken(OAuthToken oAuthToken) {
        return issuedTokens.stream()
                           .anyMatch(issuedToken -> issuedToken.getAccessToken().equals(oAuthToken.getAccessToken()));
    }

    public boolean containsRefreshToken(String refreshToken) {
        return issuedTokens.stream()
                           .anyMatch(issuedToken -> issuedToken.getRefreshToken().equals(refreshToken));
    }

    public OAuthToken getWithAccessToken(String accessToken) {
        return getFirstMatch(issuedToken -> issuedToken.getAccessToken().equals(accessToken));
    }

    public OAuthToken getWithRefreshToken(String refreshToken) {
        return getFirstMatch(issuedToken -> issuedToken.getRefreshToken().equals(refreshToken));
    }

    private OAuthToken getFirstMatch(Predicate<OAuthToken> predicate) {
        return issuedTokens.stream()
                           .filter(predicate)
                           .findFirst()
                           .get();
    }

    public boolean removeToken(OAuthToken token) {
        return issuedTokens.remove(token);
    }
}

File: src/main/scala/com/mattchowning_scala/file_read_write/server/model/OAuthTokenMap.scala
package com.mattchowning_scala.file_read_write.server.model

class OAuthTokenMap {

  private var issuedTokens: Set[OAuthToken] = Set()

  def add(oAuthToken: OAuthToken): Boolean = {
    val alreadyContainsToken: Boolean = issuedTokens.contains(oAuthToken)
    if (alreadyContainsToken) {
      System.err.println("Error: token already added to TokenMap")
      false
    } else {
      issuedTokens += oAuthToken
      true
    }
  }

  def containsAccessToken(oAuthToken: OAuthToken): Boolean =
  issuedTokens.exists(issuedToken => issuedToken.accessToken == oAuthToken.accessToken)

  def containsRefreshToken(refreshToken: String): Boolean =
    issuedTokens.exists(issuedToken => issuedToken.refreshToken == refreshToken)

  // TODO return option (i.e., drop the get, and drop the contains method)
  def getWithAccessToken(accessToken: String): OAuthToken =
    issuedTokens.find(issuedToken => issuedToken.accessToken == accessToken).get

  // TODO return option?
  def getWithRefreshToken(refreshToken: String): OAuthToken =
    issuedTokens.find(issuedToken => issuedToken.refreshToken == refreshToken).get

  def removeToken(token: OAuthToken): Boolean = {
    val containedToken = issuedTokens.contains(token)
    if (containedToken) issuedTokens -= token
    containedToken
  }
}